File: CSharp/DataStructures/Stack.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class Stack<T> : IEnumerable<T>
{

    public class Node
    {
        public T value;
        public Node next;

        public Node(T value, Node next)
        {
            this.value = value;
            this.next = next;
        }
    }

    private Node root;
    public int Size { get; private set; }

    public bool IsEmpty()
    {
        return Size <= 0;
    }

    public T Peek()
    {
        if (IsEmpty())
        {
            throw new InvalidOperationException();
        }
        return root.value;
    }

    public void Push(T value)
    {
        if (IsEmpty())
        {
            root = new Node(value, null);
            Size++;
        }
        else
        {
            Node tmp = new Node(value, root);
            root = tmp;
            Size++;
        }
    }

    public T Pop()
    {
        if (IsEmpty())
        {
            throw new InvalidOperationException();
        }
        Node tmp = root;
        root = root.next;
        Size--;
        return tmp.value;
    }

    public T Poll()
    {
        if (IsEmpty())
        {
            return default(T);
        }
        Node tmp = root;
        root = root.next;
        Size--;
        return tmp.value;
    }

    override
    public string ToString()
    {
        string res = "{";
        Node tmp = root;
        while(tmp != null)
        {
            res += tmp.value.ToString();
            if(tmp.next != null)
            {
                res += ", ";
            }
            tmp = tmp.next;
        }
        res += "}";
        return res;
    }

    public IEnumerator<T> GetEnumerator()
    {
        Node node = this.root;
        while (node != null)
        {
            yield return node.value;
            node = node.next;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return this.GetEnumerator();
    }
}

File: CPlusPlus/DataStructures/Stack.cpp
#include <stdexcept>

class Node {
    public:
        Node(int value)
            : value(value), next(nullptr)
        {}
        Node(int value, Node* next)
            : value(value), next(next)
        {}
        int value;
        Node* next;
};

class Stack {
    public:
        Stack()
            :size(0), root(nullptr)
        {}
        ~Stack();
        bool isEmpty();
        int peek();
        void push(int value);
        int pop();
    private:
        Node* root;
        int size;
};

Stack::~Stack() {
    Node* tmp = this->root;
    this->size = 0;
    while(tmp != nullptr) {
        Node* deadPtr = tmp;
        tmp = tmp->next;
        delete deadPtr;
    }
    
    this->root = nullptr;
}

bool Stack::isEmpty() {
    return size == 0;
}

int Stack::peek() {
    if(this->isEmpty()) {
        throw std::exception();
    }
    return this->root->value;
}

void Stack::push(int value) {
    if(this->isEmpty()) {
        this->root = new Node(value);
        ++this->size;
    }
    else {
        Node* tmp = new Node(value, this->root);
        this->root = tmp;
        ++this->size;
    }
}

int Stack::pop() {
    if(this->isEmpty()) {
        throw std::exception();
    }
    if(this->size == 1) {
        int tmpVal = this->root->value;
        Node* tmp = this->root;
        this->root = nullptr;
        this->size = 0;
        delete tmp;
        return tmpVal;
    }
    else {
        Node* tmp = this->root;
        int tmpVal = tmp->value;
        this->root = this->root->next;
        --this->size;
        delete tmp;
        return tmpVal;
    }
}